segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -847,6 +857,11 @@ private void defrag(MetadataTransaction txn, SegmentMetadata segmentMetadata, St
                     concatUsingAppend(concatArgs);
                 }
 
+                // Delete chunks.
+                for (int i = 1; i < chunksToConcat.size(); i++) {
+                    chunksToDelete.add(chunksToConcat.get(i).getName());
```

Comment:
what happens in case of failure ? where is recovery handled ?

Reply:
- The list of chunks to delete is returns to the caller in this case ChunkedSegmentStorage::concat
- ChunkedSegmentStorage::concat calls ChunkedSegmentStorage::collectGarbage which then immediately deletes the garbage chunks. This is called only after the whole concat operation and the metadata commit succeeds.
- What happens when delete fails ? We add it a list of "to delete list" of chunks which will be deleted by a background task.
- There is a pending issue to create background garbage collection task https://github.com/pravega/pravega/issues/4903